app/models/concerns/smart_proxy_reference.rb

```diff
@@ -0,0 +1,12 @@
+module SmartProxyReference
```

Comment:
Thanks for pointing this out. Indeed it's similar, just the query goes the other way around - in our case we start with hosts and collect smart proxy references, in hostext you start with a list of smart proxies and collect host ids.

You are right that the current code needs refactoring, since it counts on the fact that `eager_load` won't alias table names. It is true only to the point when you reference the same table twice, if you do have such reference - `eager_load` will create aliases to those tables that are not very predictable.

We can use the same concept on the smart proxy - create a set of `has_one`/`belongs_to` statements in `SmartProxy` class and mark them in the same way as we do here. If we want to force a single source of truth, we can force in the DSL to mark both ways of the connection, something along the lines of:

``` ruby
smart_proxy_reference :host => :puppet_proxy, :proxy => :puppet_host
```